File: js/sort/insertionSort.js
'use strict';
const BaseArray = require('./ArrayList.class')
class ArrayList extends BaseArray { 
  insertionSort() {
    let arrLength = this.array.length;
    for (let i = 0; i < arrLength; i++) {
      let temp = this.array[i], j = i;
      while (j > 0 && this.array[j - 1] > temp) {
        this.array[j] = this.array[j - 1]
        j--;
      }
      this.array[j] = temp;
    }
  }
}

//TestCode
let arr = new ArrayList();
arr.insert(12);
arr.insert(2);
arr.insert(7);
arr.insert(71);
arr.insert(87);
console.log(arr.toString());
arr.insertionSort();
console.log(arr.toString());

File: go-leetcodecollection/sort/insertion_sort.go
package sort

// 插入排序，a 表示数组，n 表示数组大小
func insertionSort(a []int, n int) {
	if n <= 1 {
		return
	}

	for i := 1; i < n; i++ {
		value := a[i]
		j := i - 1

		// 查找插入的位置
		for ; j >= 0; j-- {
			if a[j] > value {
				// 数据移动
				a[j+1] = a[j]
			} else {
				break
			}
		}

		// 插入数据
		a[j+1] = value
	}
}
